src/pipeline.rs

```diff
@@ -27,14 +27,67 @@ pub struct Pipeline {
 }
 
 impl Pipeline {
+    /// This is an async/await version of `Sink::send_all`, implemented because
+    /// `tokio::sync::mpsc::Sender` no longer implements `poll_ready` and therefore can't easily be
+    /// wrapped with `futures::Sink`.
```

Comment:
We might be able to use [`PollSender`](https://docs.rs/tokio-util/0.6.4/tokio_util/sync/struct.PollSender.html) here:

> A wrapper around `mpsc::Sender` that can be polled.

Reply:
Or rather, upgrade to `tokio` >= `1.1.0` and use [try_reserve](https://docs.rs/tokio/1.3.0/tokio/sync/mpsc/struct.Sender.html#method.try_reserve):

> Try to acquire a slot in the channel without waiting for the slot to become available.
>
> If the channel is full this function will return `TrySendError`, otherwise if there is a slot available it will return a `Permit` that will then allow you to `send` on the channel with a guaranteed slot. This function is similar to `reserve` except it does not await for the slot to become available.
>
> Dropping `Permit` without sending a message releases the capacity back to the channel.

Reply:
Fixed in https://github.com/timberio/vector/pull/5872/commits/aeb3db0011a0df758d408a1c2172930f1cd1bcaf. The `Permit` API also guarantees that we can't run into the case where we previously paniced, nice 😸

Reply:
Switched the whole thing to `futures::channel::mpsc` in c0666ef now. I couldn't get some of the bugs out when adapting `tokio::sync::mpsc` to a Sink API.

This should be way closer to master now, and therefore hopefully a less risky change.